tinygrad/optim.py

```diff
@@ -43,10 +43,11 @@ def __init__(self, params, lr=0.001, b1=0.9, b2=0.999, eps=1e-8):
 
   def step(self):
     self.t += 1
+    a = self.lr * (
```

Comment:
Hmm, is this really the same? pytorch doesn't seem to do this for vhat, only mhat. https://pytorch.org/docs/stable/_modules/torch/optim/adam.html#Adam

Write a test to see if it matches pytorch.

Reply:
```python
  def test_adam_step(self):
    def test_tinygrad():
      x = Tensor(x_init)
      W = Tensor(W_init)
      optim = Adam([W, x])
      m = Tensor(m_init)
      out = x.dot(W).relu()
      out = out.logsoftmax()
      out = out.mul(m).add(m).sum()
      out.backward()
      optim.step()
      return x.data, W.data

    def test_pytorch():
      x = torch.tensor(x_init, requires_grad=True)
      W = torch.tensor(W_init, requires_grad=True)
      optim = torch.optim.Adam([W, x])
      m = torch.tensor(m_init)
      out = x.matmul(W).relu()
      out = torch.nn.functional.log_softmax(out, dim=1)
      out = out.mul(m).add(m).sum()
      out.backward()
      optim.step()
      return x.detach().numpy(), W.detach().numpy()

    for x,y in zip(test_tinygrad(), test_pytorch()):
      np.testing.assert_allclose(x, y, atol=1e-5)
```
I tried to test it like this but sometimes it passes and sometimes it fails. The same thing happens with the current implementation too. 
An example output
```
AssertionError:
Not equal to tolerance rtol=1e-07, atol=1e-05

Mismatched elements: 3 / 3 (100%)
Max absolute difference: 0.00100005
Max relative difference: 0.00152729
 x: array([[ 2.761184, -0.653746,  1.145598]], dtype=float32)
 y: array([[ 2.762184, -0.654746,  1.146598]], dtype=float32)
```